src/Weeder.hs

```diff
@@ -730,22 +731,38 @@ requestEvidence n d = do
       }
 
 
--- | Follow the given evidence uses back to their instance bindings,
--- and connect the declaration to those bindings.
-followEvidenceUses :: RefMap TypeIndex -> Declaration -> Set Name -> Graph Declaration
-followEvidenceUses refMap d names =
-  let getEvidenceTrees = mapMaybe (getEvidenceTree refMap) . Set.toList
-      evidenceInfos = concatMap Tree.flatten (getEvidenceTrees names)
-      instanceEvidenceInfos = evidenceInfos & filter \case
-        EvidenceInfo _ _ _ (Just (EvInstBind _ _, ModuleScope, _)) -> True
-        _ -> False
-      evBindSiteDecls = mapMaybe (nameToDeclaration . evidenceVar) instanceEvidenceInfos
-   in star d evBindSiteDecls
+-- |Construct a map from the evidence to their declarations following the
+-- given evidence uses back to their instance bindings
+evidenceDeclarationMap :: RefMap TypeIndex -> Set Name -> Map Name [Declaration]
+evidenceDeclarationMap rf names = Set.foldl' go mempty names
```

Comment:
Using `Map.fromSet` instead of a fold should be a bit better asymptotically, and would make it explicit that the resulting map preserves the contents of `names`.

Actually - what would be even better is to change the type signature to just 
```haskell
evidenceDeclarations :: RefMap TypeIndex -> Name -> [Declaration]
```
and let `analyseEvidenceUses` build the map. Then we could use `Map.!` as mentioned in a lower comment without having to look at the implementation of `evidenceDeclarations` to justify it.

Reply:
That's how I had it initially! But then moved it around as that mostly preserved existing organization of the two functions